### ISSUE TITLE
Provide an iterator interface

### ISSUE THREAD

**Issue by zserik** [maciejhirsz/logos#67]
It would be nice if one could simply execute something like:
```rust
let toks: Vec<TokenData> = Token::lexer(source).iter().collect();
```
with a helper struct TokenData:
```rust
struct TokenData<'a, Token: Logos + Clone, RealSource<'a>> {
    token: Token,
        // = lexer.token.clone()
    slice: <RealSource<'a> as logos::Source<'a>>::Slice,
        // = lexer.slice()
    range: Range<usize>,
        // = lexer.range()
}
```

**Comment by zserik** [comment]
example: https://github.com/zserik/rustice/blob/2aae7eeae47c29635a2390bf059b2ebc58d63d8e/src/lex.rs

**Comment by edhowland** [comment]
Hi, I am new to the Logos project. Also new to Rust in general.

I haven't tried the sample code above yet, but I have need of something similar.

I am trying to wire Logos up to a Lalrpop grammar using its custom lexer facility. AFAIK it also needs an interator. It returns a 3-tuple of the form:
(usize, Tok, usize)

The first usize and the last are the start offset and end offset into the slice and the Tok is the enum varient of the match. I think I can get much of this with the lexer.range() and lexer.token method, member.

What I am strugling with is no matter how hard I try to assign the result of calling Tok::lexer("foo bar baz"); I cannot explicitly determine the type of the member.

E.g.

let mut lexer = Tok::lexer("foo bar  baz");

vs.

let mut lexer: Lexer<Tok, Source<S>> = Tok::lexer("foo bar  baz");

Obviously, the above is incorrect, but I cannot reason it out to what it should be either from compiler errors or looking at the source.
The compiler tells me that the type parameter for Source should be Slice, but from ther I get stuck. 

I really just wanted something like:

pub struct MyLexer {
  lexer: Lexer<Tok, Source<>>, /// What should go here
}

impl Iterator for MyLexer {
  pub fn next(&mut self) -> Option<(usize, Tok, usize)> {
 ... Here be majik!
}
}

Thanks for any help. 
Great work for this crate.

Regards, Ed

**Comment by maciejhirsz** [comment]
This is now implemented in #103 and will be released in 0.11:

https://github.com/maciejhirsz/logos/blob/04acab4ad7432fbb049e0c69617cc6292bd8d982/tests/tests/simple.rs#L229-L251
